### PR TITLE
Fix TSan report in StorageSet and StorageJoin

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible race condition in concurrent usage of `Set` or `Join` tables and selects from `system.tables`.

Detailed description / Documentation draft:
See https://clickhouse-test-reports.s3.yandex.net/9516/6e52fb1f5d1f751023346c80e12a357abae23be7/stress_test_(thread)/stderr.log
Found in #9516. The issue first appeared in #16306 (version 20.11).

### CODE REVIEW

File: src/Interpreters/Set.h [hunk 1]
```diff
@@ -127,8 +126,6 @@ class Set
 
     /** Protects work with the set in the functions `insertFromBlock` and `execute`.
       * These functions can be called simultaneously from different threads only when using StorageSet,
-      *  and StorageSet calls only these two functions.
```
Comment (alexey-milovidov) [inline src/Interpreters/Set.h:130]: This does not hold after #16306, hence the bug.

**Comment by alexey-milovidov** [comment]
Performance:
> 2020.12.30 18:35:19.034788 [ 93 ] {} <Trace> TLDListsHolder: public_suffix_list loading from /top_level_domains/public_suffix_list.dat
2020.12.30 18:35:19.035085 [ 93 ] {} <Error> Application: DB::ErrnoException: Cannot open file /top_level_domains/public_suffix_list.dat, errno: 2, strerror: No such file or directory
2020.12.30 18:35:19.035089 [ 93 ] {} <Information> Application: shutting down

**Comment by azat** [comment]
>2020.12.30 18:35:19.035085 [ 93 ] {} Application: DB::ErrnoException: Cannot open file /top_level_domains/public_suffix_list.dat, errno: 2, strerror: No such file or directory

I see that [performance](https://clickhouse-builds.s3.yandex.net/18385/203af8b7ec38c938185410c877e08e1a83e88061/clickhouse_build_check/report.html) package should contain it (and no errors from `cp`), however I don't see it in the `performance.tgz` archive.

**Comment by azat** [comment]
>I see that performance package should contain it (and no errors from cp), however I don't see it in the performance.tgz archive.

#18686

**Pull request merged**